source/common/router/config_impl.cc

```diff
@@ -1087,7 +1088,25 @@ RouteConstSharedPtr RouteEntryImplBase::clusterEntry(const Http::HeaderMap& head
 RouteConstSharedPtr RouteEntryImplBase::pickWeightedCluster(const Http::HeaderMap& headers,
                                                             const uint64_t random_value,
                                                             const bool ignore_overflow) const {
-  const uint64_t selected_value = random_value % total_cluster_weight_;
+  uint64_t selected_value = random_value % total_cluster_weight_;
+  // Override the selected value with provided random value if related name is specified in the
+  // header.
+  if (!random_value_name_.empty()) {
+    const auto header_value = headers.get(Envoy::Http::LowerCaseString(random_value_name_));
+    if (!header_value.empty()) {
+      ASSERT(header_value.size() == 1);
```

Comment:
I think we should probably document the behavior of multi-valued headers instead of ASSERTing this

Reply:
I have added the comment here and in the proto file to document the behavior that signal-valued header is expected and first entry is always used.

Reply:
We should remove the ASSERT and just check in the `if` above that there is only one value.

Reply:
Sure. Done